src/type.zig

```diff
@@ -567,24 +567,30 @@ pub const Type = extern union {
                 // TODO: revisit the language specification for how to evaluate equality
                 // for error set types.
 
-                if (a.tag() == .anyerror and b.tag() == .anyerror) {
-                    return true;
-                }
-
-                if (a.tag() == .error_set and b.tag() == .error_set) {
-                    return a.castTag(.error_set).?.data.owner_decl == b.castTag(.error_set).?.data.owner_decl;
-                }
+                if (a.tag() != b.tag()) return false;
```

Comment:
this looks wrong - why can't the tags be different?

Reply:
This is the same logic as before (see split diff!)

Reply:
the `return false` at the end of master branch is wrong - it's essentially returning a false negative as a TODO.

Reply:
I can turn this into `unreachable` for now if that'll unblock this PR, since it feels like merging the changes of #11022 into this might be heavy (I have those in a branch, but requires more changes i.e. due to resolving inferred error sets).

Reply:
Alright, #11022 is marked "accepted" now, hope that helps :-)

I'm fine with however you want to structure the changesets.

Also I think we can probably get away with implementing this proposal only in stage2 and leaving stage1 to rot for now. In case you haven't seen it, this is the plan to deal with that: #6378

Reply:
Hah okay, stand by. I'll re-request review once those are in this PR.

Reply:
Added it, note that even though the changes are only in stage2, I don't think anything was or is exercising the changed behavior in the compiler (besides new tests). So in regards to #6378 I don't think this will negatively impact bootstrapping in any way.